core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -241,7 +241,19 @@ class LogCleaner(val config: CleanerConfig,
           } finally {
             cleanerManager.doneCleaning(cleanable.topicPartition, cleanable.log.dir.getParentFile, endOffset)
           }
+          true
       }
+      val deleted: Iterable[(TopicAndPartition, Log)] = cleanerManager.logsWithSegmentsReadyToBeDeleted()
+      deleted.foreach{
+        case (topicPartition, log) =>
+          try {
+            log.deleteOldSegments()
+          } finally {
+            cleanerManager.doneDeleting(topicPartition)
+          }
+      }
+      if (!cleaned && deleted.isEmpty)
```

Comment:
If there is a topic with compact and delete, deleted will always be non-empty. So, perhaps we can just check cleaned?